### PR TITLE
New Backend API

### PR DESCRIPTION
**Pull request by rkimballn1**
Motivations:
1) Currently we use `shared_ptr<Function>` as the handle for compile and all calls. `shared_ptr<Function>` is mutable and can change from call to call, so assuming it is constant is bad.
2) Currently we pass `shared_ptr<Function>` to compile and when we return we have modified `Function` through compile. If you try to compile the same `Function` twice or try to use the same `Function` on two different backends you get an error or worse. The ultimate goal is to pass a `const Function&` to compile but that is a follow-on PR. Some frameworks use the side-effect of the modified `Function` to get the function's input/output shapes. Two new functions `get_parameters` and `get_results` are added to give a more robust way of getting this information.
3) Using an opaque `Handle` as the handle for all backend calls allows each backend to use the `Handle` for any internal purpose.
4) A follow-on PR will add `Save` and `Load` functions to save/load a compile function. It makes sense to have `Load` return a `Handle` instead of a `shared_ptr<Function>` since the backend may have no use for the `Function` and the current serializer can't serialize/deserialize compiled `Function`s

~~Everything except for some deprecated functions now use Handle instead of shared_ptr<Function>. Handle is an opaque pointer backends can use for whatever they want. This PR uses Handle to point to the backends FunctionInstance.~~
runtime::Backend::compile now returns a `unique_ptr<runtime::Executable>`. This Executable object contains all of the methods which used to take `Handle`. There is no caching of compiles. An Executable object lives as long as the user maintains a reference to it and no longer. Thank you @earhart for this suggestion.

runtime::Backend ends up containing only `create_tensor` and `compile`.

A new function `execute` replaces `call` 
A new function `validate_and_execute` replaces `call_with_validate`. This function is only used in unit tests so it should eventually migrate there.

`compile` will be changed in a future PR to take a `const Function&` rather than a shared_ptr<Function>. As a const it will not be modified during compile. There are two new functions `get_parameters` and `get_results` that get the Parameters and Results for the compiled function. These can be used to create bprop.

### CODE REVIEW

**Comment by earhart** [comment]
I'm not sure why `Handle` is a `void*` -- it requires a non-standard destructor (more fragile & prone to error than the standard destructor pattern), it forces the backend to use a `static_cast` to get to its function instance (no possibility for error checking -- unless the backend wants to implement a handle indirection table, anyway), and it requires that users keep a `Backend` instance around in addition to the `Handle`.

Since we're already requiring that the interlibrary ABI permits shared C++ types, why not replace `Handle` with an abstract base that provides `execute()`, `validate()`, &c (replacing the backend methods that're taking a `Handle`) and have `compile()` return a `std::unique_ptr<>` to the abstract base?  And then the backend can find out when the function's freed by observing the destructor invocation, downcasts happen automatically, and the caller doesn't have to even think about which `Handle` instances go with which `Backend`.

**Comment by rkimballn1** [comment]
@earhart I like that idea! The current PR is doing things is a very C way while your suggestion is much more C++. I will make the changes and see what happens.

File: src/ngraph/runtime/executable.hpp [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+    /// \returns an ngraph::ResultVector of all input parameters
+    const ngraph::ResultVector& get_results() const;
+
+protected:
+    /// \brief Called at the end of compile to the the values to be returned by get_parameters
+    ///     and get_results
+    /// \param func The function with Results fully resolved.
+    void set_parameters_and_results(const Function& func);
+
+    runtime::Backend* get_backend();
```
Comment (earhart) [inline src/ngraph/runtime/executable.hpp]: Just curious -- why is this here?  (It doesn't seem to be used by the base class; it seems reasonable to let the subclass track it iff the subclass needs it -- and then the subclass can get the type right.)

File: src/ngraph/runtime/executable.cpp [hunk 2]
```diff
@@ -0,0 +1,153 @@
+//*****************************************************************************
+// Copyright 2017-2018 Intel Corporation
```
[the remark this answers is not in the document]
Comment (rkimballn1) [reply src/ngraph/runtime/executable.cpp]: yes

File: src/ngraph/runtime/backend.hpp [hunk 3]
```diff
@@ -31,14 +31,16 @@ namespace ngraph
...
+        using Handle = std::unique_ptr<Executable>;
+        using SharedHandle = std::shared_ptr<Executable>;
     }
 }
 
 /// \brief Interface to a generic backend.
 ///
 /// Backends are responsible for function execution and value allocation.
-class ngraph::runtime::Backend
+class ngraph::runtime::Backend : public std::enable_shared_from_this<Backend>
```
[the remark this answers is not in the document]
Comment (rkimballn1) [reply src/ngraph/runtime/backend.hpp]: This is a leftover of a failed attempt. Removed

File: src/ngraph/runtime/executable.hpp [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+
+    /// \brief Query the output Results for a given Handle
+    /// \returns an ngraph::ResultVector of all input parameters
+    const ngraph::ResultVector& get_results() const;
+
+protected:
+    /// \brief Called at the end of compile to the the values to be returned by get_parameters
+    ///     and get_results
+    /// \param func The function with Results fully resolved.
+    void set_parameters_and_results(const Function& func);
```
[the remark this answers is not in the document]
Comment (rkimballn1) [reply src/ngraph/runtime/executable.hpp]: They must be set post compile. The type propagation and layout needs to be run so that the Results are valid for mxnet to produce bprop.

File: src/ngraph/runtime/executable.hpp (hunk 1 above)
Comment (rkimballn1) [reply src/ngraph/runtime/executable.hpp]: There are some ops like FunctionCall that need to compile graphs. Since these ops are handled in the Executable I think I can remove this since the Executable is the object that does the compile.
It is currently only used in the INTERPRETER backend.

Comment (earhart) [reply src/ngraph/runtime/executable.hpp]: <nods> In the PlaidML backend, there's a Build object that contains a pointer to a Compiler object, for exactly this sort of thing.

Comment (rkimballn1) [reply src/ngraph/runtime/executable.hpp]: I removed the need for get_backend() since the interpreter can just make a new executable internally when it is needed for ops like FunctionCall.

File: src/ngraph/runtime/executable.hpp [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+    }
+}
+
+class ngraph::runtime::Executable
+{
+public:
+    Executable(runtime::Backend* backend);
+    virtual ~Executable();
+
+    /// \deprecated use execute method
```
[the remark this answers is not in the document]
Comment (rkimballn1) [reply src/ngraph/runtime/executable.hpp]: The PR is already Godzilla sized. I did this so not every like of code has to change quite yet.

File: src/ngraph/runtime/executable.cpp [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+
+void runtime::Executable::set_parameters_and_results(const Function& func)
+{
+    m_parameters = func.get_parameters();
+    m_results = func.get_results();
+}
+
+vector<runtime::PerformanceCounter> runtime::Executable::get_performance_data() const
+{
+    return vector<PerformanceCounter>();
```
[the remark this answers is not in the document]
Comment (rkimballn1) [reply src/ngraph/runtime/executable.cpp]: changed

File: src/ngraph/runtime/executable.hpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (mbrookhart) [reply src/ngraph/runtime/executable.hpp]: CPU compilation of the forward function changes the layout of the output tensors, we need to be able to grab that information post-compilation to set up the backward function.

File: src/ngraph/runtime/intelgpu/CMakeLists.txt [hunk 7]
```diff
@@ -16,6 +16,7 @@
 
 set(SRC
     intelgpu_backend.cpp
+    intelgpu_executable.cpp
```
Comment (shssf) [inline src/ngraph/runtime/intelgpu/CMakeLists.txt]: @rkimballn1 Is it possible to make all required changes in the same files? I mean, moving code into other file looks like a bit diffrent thing than this PR. I'm afraid some modern changes in the Backend might be lost because they are invisible due to function moved into new file.
